solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol

```diff
@@ -0,0 +1,1386 @@
+     * note that the owner needs to call acceptOwnership on the new converter after the upgrade
+     */
+    function upgrade() public ownerOnly {
+        IConverterUpgrader converterUpgrader = IConverterUpgrader(addressOf(CONVERTER_UPGRADER));
+
+        // trigger de-activation event
+        emit Activation(converterType(), anchor, false);
+
+        transferOwnership(address(converterUpgrader));
+        converterUpgrader.upgrade(version);
+        acceptOwnership();
+    }
+
+    /**
+     * @dev returns the number of reserve tokens
+     * note that prior to version 17, you should use 'connectorTokenCount' instead
+     *
+     * @return number of reserve tokens
+     */
+    function reserveTokenCount() public view returns (uint16) {
+        return uint16(__reserveTokens.length);
+    }
+
+    /**
+     * @dev returns the array of reserve tokens
+     *
+     * @return array of reserve tokens
+     */
+    function reserveTokens() public view returns (IERC20Token[] memory) {
+        return __reserveTokens;
+    }
+
+    /**
+     * @dev defines a new reserve token for the converter
+     * can only be called by the owner while the converter is inactive
+     *
+     * @param _token   address of the reserve token
+     * @param _weight  reserve weight, represented in ppm, 1-1000000
+     */
+    function addReserve(IERC20Token _token, uint32 _weight)
+        public
+        virtual
+        override
+        ownerOnly
+        inactive
+        validAddress(address(_token))
+        notThis(address(_token))
+        validReserveWeight(_weight)
+    {
+        // validate input
+        require(address(_token) != address(anchor) && __reserveIds[_token] == 0, "ERR_INVALID_RESERVE");
+        require(reserveTokenCount() < 2, "ERR_INVALID_RESERVE_COUNT");
+
+        __reserveTokens.push(_token);
+        __reserveIds[_token] = __reserveTokens.length;
+    }
+
+    /**
+     * @dev returns the reserve's weight
+     * added in version 28
+     *
+     * @param _reserveToken    reserve token contract address
+     *
+     * @return reserve weight
+     */
+    function reserveWeight(IERC20Token _reserveToken) public view validReserve(_reserveToken) returns (uint32) {
+        return PPM_RESOLUTION / 2;
+    }
+
+    /**
+     * @dev returns the balance of a given reserve token
+     *
+     * @param _reserveToken    reserve token contract address
+     *
+     * @return the balance of the given reserve token
+     */
+    function reserveBalance(IERC20Token _reserveToken) public view override returns (uint256) {
+        uint256 reserveId = __reserveIds[_reserveToken];
+        require(reserveId != 0, "ERR_INVALID_RESERVE");
+        return reserveBalance(reserveId);
+    }
+
+    /**
+     * @dev returns the balances of both reserve tokens
+     *
+     * @return the balances of both reserve tokens
+     */
+    function reserveBalances() public view returns (uint256, uint256) {
+        return reserveBalances(1, 2);
```

Comment:
Maybe 0 and 1 instead?

Reply:
With `mapping(IERC20Token => uint256) private __reserveIds`, we are assuming that our two reserve tokens are mapped to 1 and 2, and all other possible addresses are mapped to 0.